firedrake/ufl_expr.py

```diff
@@ -138,11 +139,29 @@ def derivative(form, u, du=None, coefficient_derivatives=None):
     See also :func:`ufl.derivative`.
     """
     # TODO: What about Constant?
-    if len(u.split()) > 1 and set(extract_coefficients(form)) & set(u.split()):
+    notx = not isinstance(u, SpatialCoordinate)
+    if notx and len(u.split()) > 1 and set(extract_coefficients(form)) & set(u.split()):
         raise ValueError("Taking derivative of form wrt u, but form contains coefficients from u.split()."
                          "\nYou probably meant to write split(u) when defining your form.")
     if du is None:
-        if isinstance(u, firedrake.Function):
+        u_is_spatialcoordinate = isinstance(u, ufl.SpatialCoordinate)
+        u_is_mesh_coordinate = u is form.ufl_domain().coordinates
+
+        if u_is_spatialcoordinate or u_is_mesh_coordinate:
+            if u_is_mesh_coordinate:
+                coords = u
+                u = SpatialCoordinate(u.ufl_domain())
+            else:
+                coords = u.ufl_domain().coordinates
+            V = coords.function_space()
+            args = form.arguments()
+            number = max(a.number() for a in args) if args else -1
+            du = Argument(V, number + 1)
+            cds = {coords: du}
+            if coefficient_derivatives is not None:
+                cds.update(coefficient_derivatives)
+            coefficient_derivatives = cds
+        elif isinstance(u, firedrake.Function):
```

Comment:
Sorry for letting this languish so long.

I think we need to provide the coefficient_derivatives in all cases where a coordinate derivative is taken.

How about this logic?

```python
def derivative(f, u, du=None, coefficient_derivatives=None):
   ...
   is_dX = isinstance(u, SpatialCoordinate) or u is f.ufl_domain().coordinates

   def argument(V):
       n = max(a.number() for a in args) if args else -1
       return Argument(V, n + 1)

   if is_dX:
       coords = f.ufl_domain().coordinates
       u = SpatialCoordinate(f.ufl_domain())
       V = coords.function_space()
       if du is None:
           du = argument(V)
       cds = {coords: du}
       if coefficient_derivatives is not None:
           cds.update(coefficient_derivatives)
       coefficient_derivatives = cds
    elif isinstance (u, ...):
        ...
```

WDYT?

Reply:
You're absolutely right, fixed it.